crates/driver/src/domain/competition/mod.rs

```diff
@@ -390,6 +477,196 @@ impl Competition {
+                    auction_id = %pod_auction_id,
+                    deadline = %deadline,
+                    "[pod] bid submission failed"
+                );
+                return Err(e);
+            }
+        }
+
+        Ok(())
+    }
+
+    pub async fn pod_fetch_participants(
+        pod_auction_client: &AuctionClient,
+        auction: &Auction,
+        deadline: chrono::DateTime<chrono::Utc>,
+    ) -> Result<Vec<ParticipantDetached<Unranked>>, anyhow::Error> {
+        let pod_auction_id = match u64::try_from(auction.id().map(|id| id.0).unwrap_or_default()) {
+            Ok(id) => pod_sdk::U256::from(id),
+            Err(e) => {
+                tracing::error!(error = %e, "[pod] invalid auction id");
+                return Err(anyhow::Error::new(e));
+            }
+        };
+
+        if let Err(e) = pod_auction_client
+            .wait_for_auction_end(deadline.into())
+            .await
+        {
+            tracing::error!(
+                error = %e,
+                auction_id = %pod_auction_id,
+                deadline = %deadline,
+                "[pod] wait_for_auction_end failed"
+            );
+            return Err(e);
+        }
+        tracing::info!(
+            auction_id = %pod_auction_id,
+            deadline = %deadline,
+            "[pod] auction ended"
+        );
+
+        let bids = match pod_auction_client.fetch_bids(pod_auction_id).await {
+            Ok(bids) => bids,
+            Err(e) => {
+                tracing::error!(
+                    error = %e,
+                    auction_id = %pod_auction_id,
+                    deadline = %deadline,
+                    "[pod] fetch bids failed"
+                );
+                return Err(e);
+            }
+        };
+
+        let mut participants = Vec::new();
+        for bid in bids {
+            let bidder_solve_response: dto::SolveResponse =
+                match serde_json::from_slice(bid.data.as_slice()) {
+                    Ok(resp) => resp,
+                    Err(e) => {
+                        tracing::error!(
+                            error = %e,
+                            auction_id = %pod_auction_id,
+                            deadline = %deadline,
+                            "[pod] failed to deserialize SolveResponse"
+                        );
+                        return Err(anyhow::Error::new(e));
+                    }
+                };
+
+            for solution in bidder_solve_response.solutions {
+                participants.push(ParticipantDetached::new(solution));
+            }
+        }
+
+        Ok(participants)
+    }
+
+    pub async fn local_winner_selection(
+        auction: &Auction,
+        participants: Vec<ParticipantDetached<Unranked>>,
+        arbitrator: &LocalArbitrator,
+    ) -> Result<(), anyhow::Error> {
+        let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
+            arbitrator.arbitrate(participants, auction)
+        }));
```

Comment:
Is there any specific panic that happens here that caused you to add this catch_unwind?